# Tutorial for data preprocessing

---
title: "Data Preprocessing"
author: "Yang Liao"
date: "2023-09-02"
output:
  html_document: default
  pdf_document: default
---

In DMLA cycle, 'D' represents discover, i.e. experimental discovery of special biological response that can be captured by meta-omics datasets. 'M' represents model, i.e. utilizing biological domain-knowledge integrated models to characterize complex biological process.

In this tutorial, we demonstrate the preprocess of raw datasets, i.e. the 'D' stage in DMLA cycle, to obtain basic information and datasets for geometric deep leanring, i.e. the 'M' stage.

We start by define the dataset name, as well as paths for expression levels/abundance and annotations. After that, we conduct DESeq analysis to obtain DEGs (Differentially Expressed Genes) through obtain_DEGs(). After that, we obtain the datasets (dgi dataset) for geometric deep learning and graph data based on TSNE analysis, i.e. get_tsne_df(). This function returns graph_data_list, a list that contain graph datasets and dgi datasets.

The following DEMOs are based on different datasets.

## DEMO1: Optogenetic denitrifier based on metatranscriptomics

### 'D': Discover

-   The original cell output is as follows:

```         
[1] "The total number of DEGs (not unique): 109276"
[1] "The total number of unique DEGs (not unique): 55189"
[1] "The total number of deseq filtered DEGs: 42666"
[1] "The total number of unique deseq filtered DEGs: 37217"
[1] "The number of total genes (unique): 56991"
[1] "The number of total DEGs after filtering : 29643"

  Blue_vs_Dark Yellow_vs_Dark 
         25277           4366 
[1] "The total number of unique DEGs for tsne analysis of Blue_vs_Dark dataset is: 25277"
[1] "The total number of unique DEGs for tsne analysis of Yellow_vs_Dark dataset is: 4366"
```

```{r DEMO1}
rm(list=ls()) 

# load function
source('./scr/R/PreprocessingFunctions.R')

# === import data path ===
# --- DESeq ---
dataset <- 'LY_9samples_metatranscriptomics'
rawdata_path <- paste("./data/", dataset, "/rawdata/reads_number.txt", sep = '')
group_path <- paste("./data/", dataset, "/inputdata/group.csv", sep = '')
output_path <- paste("./data/", dataset, "/inputdata/", sep = '') # input data for modeling in python
output_path.rawDEGs <- paste(output_path, 'DEGs_raw.csv', sep = '')

# --- filtered DEGs ---
rpkm_path <- paste("./data/", dataset, "/rawdata/RPKM.txt", sep = '')
DEGs.filtered.path <- paste("./data/", dataset, "/inputdata/DEGs_filtered.csv", sep = '')

## subcellular annotation
anno.signal <- paste("./data/", dataset, "/rawdata/Signal.csv", sep = '')
anno.tmhmm <- paste("./data/", dataset, "/rawdata/tmhmm.csv", sep = '')

# === conduct analysis ===
# obtain DEGs
DEGs.filtered <- obtain_DEGs(rawdata_path, group_path, 
                        rpkm_path, thre = 1,
                        output_path.rawDEGs, DEGs.filtered.path)

# obtain graph data
graph_data_list <- get_tsne_df(DEGs.filtered, dataset,
                               group_path,
                               anno.info.1=anno.signal,
                               anno.info.2=anno.tmhmm)

```

Here is a visualization toul for dataset exploratory analysis. For example, the distribution of subcellular information among all genes.

```{r Visualization}
# ========= visualization toolkits =========

# ---  tsne plot ---
graph_data <- graph_data_list$Blue_vs_Dark
ggplot(graph_data,aes(tSNE1,tSNE2, colour = as.factor(tmhmm))) +  # ,colour = degs
  geom_point(alpha = 0.5, size = 1) + 
  labs(title = '') + 
  theme_bw() + 
  scale_colour_manual(values = pale_25) +
  mytheme1
```

### 'M': Model

In this part, we utilized geomteric deep learning to unsupervised learned the genetic co-expression, i.e. obtaining gene panels.

```{python}
from scr.config import dataset_name, group, config
dataset_name = 'LY_15samples_metagenomics'
group = 'Yellow_vs_Dark'
config['threshold'] = 0.4
config['epoch'] = 20000
```

```{bash}
cd ./scr
source /Users/yangliao/opt/anaconda3/bin/activate pyg
# conda activate pyg
python3 main.py
```

## DEMO2: Optogenetic anaerobic CO2 fixation for denitrification based on metatranscriptomics

## DEMO3: Optogenetic denitrifier based on metagenomics

-   The original cell output is as follows:

```         
[1] "The total number of DEGs (not unique): 788036"
[1] "The total number of unique DEGs (not unique): 197009"
[1] "The total number of deseq filtered DEGs: 15157"
[1] "The total number of unique deseq filtered DEGs: 12021"
[1] "The number of total genes (unique): 198113"
[1] "The number of total DEGs after filtering : 922"

  Blue_vs_Dark  Green_vs_Dark    Red_vs_Dark Yellow_vs_Dark 
            88            135            642             57 
[1] "The total number of unique DEGs for tsne analysis of Blue_vs_Dark dataset is: 88"
[1] "The total number of unique DEGs for tsne analysis of Green_vs_Dark dataset is: 135"
[1] "The total number of unique DEGs for tsne analysis of Red_vs_Dark dataset is: 642"
[1] "The total number of unique DEGs for tsne analysis of Yellow_vs_Dark dataset is: 57"
```

```{r}
rm(list=ls()) 
source('./scr/R/PreprocessingFunctions.R')

# === import data path ===
# --- DESeq ---
dataset <- 'LY_15samples_metagenomics'
rawdata_path <- paste("./data/", dataset, "/rawdata/reads_number.txt", sep = '')
group_path <- paste("./data/", dataset, "/inputdata/group.csv", sep = '')
output_path <- paste("./data/", dataset, "/inputdata/", sep = '') # input data for modeling in python
output_path.rawDEGs <- paste(output_path, 'DEGs_raw.csv', sep = '')

# --- filtered DEGs ---
rpkm_path <- paste("./data/", dataset, "/rawdata/RPKM.txt", sep = '')
DEGs.filtered.path <- paste("./data/", dataset, "/inputdata/DEGs_filtered.csv", sep = '')

## subcellular annotation
anno.signal <- paste("./data/", dataset, "/rawdata/Signal.csv", sep = '')
anno.tmhmm <- paste("./data/", dataset, "/rawdata/tmhmm.csv", sep = '')

# === conduct analysis ===
# obtain DEGs
DEGs.filtered <- obtain_DEGs(rawdata_path, group_path, 
                             rpkm_path, thre = 1, 
                             output_path.rawDEGs,
                             DEGs.filtered.path)
# obtain graph data
graph_data_list <- get_tsne_df(DEGs.filtered, dataset,
                               group_path, perplex = 10)
```

## DEMO4: Optogenetic anaerobic CO2 fixation for denitrification based on metagenomics
