## APP4: Topology Network with landmark genes regulatory strategy

```{r dataloader}
# setting parameters
dataset = 'LY_9samples_metatranscriptomics'
group = 'Yellow_vs_Dark'
SubDataset <- 'YvsD_SubCell_dgi'
Cluster = 'Cluster 3' # HGP

# labeling gene panels
plotdata_cluster.list <- list()
i = 1
# Cluster <- paste('Cluster', i)
cluster_path <- paste("./data/", dataset, "/output_data/R/", SubDataset, '_', Cluster, '_Enrichment.csv', sep = '') 
Hub_yellow_raw <- read.csv(cluster_path)

# ko descrption is kept as labels
kegg_anno.path <- paste("./data/", dataset, "/rawdata/kegg_annotation_all.csv", sep = '') 
kegg_anno <- read.csv(kegg_anno.path)

# --- obtain pathway list & matching & obtain gene list ---
path_list <- Func_YvsD$level3_pathway_name

# initial
path <- path_list[1]
# for loop 
gene_list <- rawdata[grepl(path, rawdata$level3_pathway_name), ] 
for (path in path_list[2:length(path_list)]){
  list <- rawdata[grepl(path, rawdata$level3_pathway_name), ]
  gene_list <- rbind(gene_list, list)
}


# merge data: id, samples expression levels, subcellular info, ko_des
ko <- c('id', 'ko', 'ko_name', 'ko_des', 'level3_pathway_name')
Hub_yellow <- merge(Hub_yellow_raw, kegg_anno[ko], by = 'level3_pathway_name') # Hub_yellow_row: 没有id, 可能需要根据其他的去merge —— 再看看之前pathway & gene之间如何协调的 (能否把id找出来，不然现在的肯定不对（根据pathway去match可能多match了）)

# === load data v2: gene interaction - split datasets, yellow vs dark ===
# hub
Hub_yellow_raw <- read.csv('./03_DGI/output_data/GeneListofFilteredPathway_YvsD_Cluster3.csv', header = TRUE) # Corresponds to enrichment analysis output, but annotation need to be matched

Hub_yellow <- Hub_yellow_raw[, c(1, 13:21, 9:10, (length(Hub_yellow_raw) -1))]
merge_data <- unique(Hub_yellow)
# signaling
signal_yellow_raw <- read.csv('./03_DGI/output_data/GeneListofFilteredPathway_YvsD_SignalCluster1.csv', header = TRUE)
signal_yellow <- signal_yellow_raw[, c(1, 13:21, 9:10, (length(signal_yellow_raw) -1) )]
phototransduction_yellow_raw <- read.csv('./03_DGI/output_data/ClusterAssign_phototransduction_YvsD_7Cluster.csv', header = T)
phototransduction_yellow <- phototransduction_yellow_raw[, c(1, 13:21, 9:10, (length(phototransduction_yellow_raw) -1) )]
merge_data <- unique(rbind(signal_yellow, phototransduction_yellow))


```

From app3, we can identify the Hub Gene Panels (HGPs) and Signaling Gene Panels (HGPs).

-   HGP

-   SGP

To better regulate the microbiome, we construct regulatory network, also called interaction network, on the genes level based on microbial genetic topology.

-   ObtainGP(): Including HGP, SGP, for further process [to be done]

```{r ObtainGP()}
# setting parameters
dataset = 'LY_9samples_metatranscriptomics'
group = 'Yellow_vs_Dark'
SubDataset <- 'YvsD_SubCell_dgi'
Cluster = 'Cluster 3' # HGP


# labeling gene panels
plotdata_cluster.list <- list()
i = 1
# Cluster <- paste('Cluster', i)
cluster_path <- paste("./data/", dataset, "/output_data/R/", SubDataset, '_', Cluster, '.csv', sep = '') 


HGP_y <- unique(read.csv(cluster_path))
HGP_y$expre <- apply(HGP_y[,7:9], 1, mean)
```
